README.md
## 🌐 NET-CAT

TCPChat is a simple TCP-based chat server implemented in Go. It allows multiple clients to connect and communicate with each other in a real-time chat environment. The server keeps track of connected users, broadcasts messages to all connected users, and logs the chat history to a file.

## DESCRIPTION
The TCPChat project is a simple but effective implementation of a TCP-based chat server in Go. It allows users to connect, exchange messages, and see real-time updates from other connected users. The server provides a straightforward command-line interface and handles multiple connections concurrently using goroutines.

The project includes the following key features:

- **Real-time Communication**: Users can send and receive messages in real-time, creating an interactive chat experience.

- **User Management**: The server keeps track of connected users, ensuring unique names and limiting the number of simultaneous connections to 10.

- **Chat History Logging**: The server logs all chat messages to a file, enabling future reference and analysis of the conversation.

The TCPChat project can serve as a foundation for building more advanced chat applications or as a learning resource for understanding network programming concepts using Go.

We hope you find this project useful and enjoy exploring the world of TCP-based chat applications with TCPChat!

## USAGE

1. Download or clone the repository to your local machine.

2. Navigate to the project directory.

3. Build the server executable using the following command:
   ```
   go build .
   ```

4. Run the server using the following command:
   ```
   ./netcat
   ```

5. The server will start listening for incoming connections on the default port 8989. If you want to use a different port, you can specify it as a command-line argument:
   ```
   ./netcat 8080
   ```

6. Clients can connect to the server using Netcat client. For example, using Telnet:
   ```
   nc localhost 8989
   ```

7. Once connected, clients can start sending and receiving messages. Type your name to join the chat room.

## AVAILABLE COMMANDS

1. To clear the screen:
   ```
   /clear
   ```

2. To list all connected users:
   ```
   /list
   ```

3. To send a private message to a specific user:
   ```
   @username Your message here
   ```

4. To exit the chat:
   ```
   /exit
   ```

## NOTE

- The chat server has a maximum limit of 10 users. If the server is full, new connections will be rejected.

- The server logs all chat messages to a log file named "chat.log" in the "log" directory.

- Emoji shortcuts in messages will be automatically replaced with their corresponding emoji characters.

- Please ensure that you have Telnet or Netcat installed on your machine to connect to the server from the command line.

##  AUTHOR
+   Cheikh Ahmadou Tidiane Diallo
+   Serigne Saliou Mbacké Mbaye
+   Pape Bilaly Sow


data/history.txt
qaz has joined our chat... 
wsx has joined our chat... 
sdf has joined our chat... 


test/main.go
package main

import (
	"fmt"
	"net"
	"os"
	"time"
)

func main() {
	var conn net.Conn
	var err error
	// Retry until connection is established
	for {
		conn, err = net.Dial("tcp", "11.11.90.185:2525")
		if err == nil {
			break
		}
		fmt.Println("❌ Failed to connect, retrying in 10 seconds:", err)
		time.Sleep(10 * time.Second)
	}
	defer conn.Close()

	fmt.Println("✅ Connection established")
	
	fmt.Fprintln(conn, "BotBot") // Sending the bot name to the server.
	
	fmt.Println("😂 Sending messages...")
	
	for i := 0; i < 100000; i++ { // Send 10000 messages.
		_, err := fmt.Fprintf(conn, "👋 Hello from BotBot %d\n", i)
		if err != nil {
			fmt.Println("Failed to send message:", err)
			os.Exit(1)
		}
		time.Sleep(1 * time.Millisecond) // Wait for 1 second before sending the next message.
	}

	fmt.Println("DONE.")
	// for i := 45; i < 254; i++ {
	// 	ip := fmt.Sprintf("11.11.90.%d:8989", i)
	// 	fmt.Println("Try to connect to:", ip)
	// 	conn, err := net.Dial("tcp", ip)
	// 	if err != nil {
	// 		fmt.Println("Failed to connect to ", ip, " : ", err)
	// 		// os.Exit(1)
	// 		continue
	// 	}
	// 	fmt.Println("Connection established with the ip:", ip)
	// 	defer conn.Close()

	// 	fmt.Fprintln(conn, "Bot-", i) // Sending the bot name to the server.

	// 	for i := 0; i < 10000; i++ { // Send 1000 messages.
	// 		_, err := fmt.Fprintf(conn, "This is message number %d\n", i)
	// 		if err != nil {
	// 			fmt.Println("Failed to send message:", err)
	// 			os.Exit(1)
	// 		}
	// 		time.Sleep(1 * time.Millisecond) // Wait for 1 millisecond before sending the next message.
	// 	}
	// }
}


main.go
package main

import (
	"flag"
	"fmt"
	"net"
	netcat "netcat/lib"
	"os"
	"time"
)

func main() {
	args := os.Args[1:]
	flag.IntVar(&netcat.MaxConnections, "m", 10, "Maximum number of concurrent connections allowed")
	flag.Parse()
	if len(args) > 1 {
		fmt.Println("❌ [USAGE]: ./TCPChat $port")
	} else {
		port := "8989"
		if len(args) != 0 {
			port = args[0]
		}
		listener, err := net.Listen("tcp", ":"+port)
		if err != nil {
			fmt.Println("Failed to launch server: ", err)
		}
		fmt.Printf("🚀 Server listening on the port :%s\n", port)

		timestamp := time.Now()
		fileName := fmt.Sprintf("logs/chat-log-%s.log", timestamp)
		netcat.LogFile, err = os.Create(fileName)
		if err != nil {
			fmt.Println("Failed to create log file: ", err)
		}
		for {
			connection, _ := listener.Accept()
			go netcat.Chat(connection)
		}
	}
}


lib/chat.go
package netcat

import (
	"bufio"
	"fmt"
	"net"
	"os"
	"strings"
	"time"
)

// Chat handles the communication with a new user connecting to the chat.
// It checks if the chat is full, if not, it welcomes the user and asks for their name.
// If the name is already in use, it informs the user and asks for a different name.
// Once the user is added to the chat, they can participate in the conversation.
//
// Parameters:
//
//	user (net.Conn): The network connection representing the user.
//
// Goals:
//  1. Handle user registration and name validation.
//  2. Manage user interactions in the chat, such as sending messages and commands.
//  3. Handle special commands like renaming, private messages, listing users, and clearing the screen.
//  4. Broadcast messages to all other users in the chat.
//  5. Handle user disconnection and remove them from the chat.
func Chat(user net.Conn) {
	if len(users) >= MaxConnections {
		user.Write([]byte("🔒 Sorry chat is full. Try again later\n"))
		user.Close()
		return
	}
	var name, message string
	welcome, err := os.ReadFile("assets/logo.txt")
	if err != nil {
		fmt.Println("❌ [ERROR]: Cannot open file\n" + err.Error())
	}
	user.Write(welcome)
	scanner := bufio.NewScanner(user)
	for scanner.Scan() {
		name = scanner.Text()
		name = strings.TrimSpace(name)
		_user, isNameAlreadyUse := users[name]
		if len(name) == 0 {
			user.Write([]byte("❌ Please enter correct name.\n[ENTER YOUR NAME]:"))
		} else if isNameAlreadyUse {
			user.Write([]byte("❌ The name is already use in the room.\n[ENTER YOUR NAME]:"))
			_user.Write([]byte("\r\033[K ❌ Someone try to use your logging name\n"))
			_user.Write([]byte(fmt.Sprintf("[%s][%s]:", time.Now().Format("2006-01-02 15:04:05"), name)))
		} else {
			users[name] = user
			break
		}
	}

	if name != "" {
		user.Write([]byte(messages))
		broadcast(fmt.Sprintf("🤝 %s has joined our chat ...\n", name), name, false)
		for {
			ok := scanner.Scan()
			if !ok {
				break
			}
			text := replaceEmojis(scanner.Text())
			if len(strings.Trim(text, " \n\t\r")) != 0 {
				if strings.HasPrefix(text, ">rename ") {
					newName := strings.TrimSpace(text[8:])
					if newName == "" {
						user.Write([]byte("❌ Invalid name.\n"))
						user.Write([]byte(fmt.Sprintf("[%s][%s]:", time.Now().Format("2006-01-02 15:04:05"), name)))
					} else if _, ok := users[newName]; ok {
						user.Write([]byte("❌ Name is already in use.\n"))
						user.Write([]byte(fmt.Sprintf("[%s][%s]:", time.Now().Format("2006-01-02 15:04:05"), name)))
					} else {
						delete(users, name)
						prevName := name
						name = newName
						users[name] = user
						user.Write([]byte("✅ Name changed successfully.\n"))
						broadcast(fmt.Sprintf("🗣️  %s has change his name to %s ...\n", prevName, name), name, false)
					}
				} else if strings.HasPrefix(text, "@") {
					_text := strings.Split(text, " ")
					private := string(_text[0][1:])
					message = fmt.Sprintf("@[%s][%s]:", time.Now().Format("2006-01-02 15:04:05"), name) + strings.Join(_text[1:], " ") + "\n"
					_user, err := users[private]
					if !err {
						user.Write([]byte("❌ User not exit.\n"))
					} else {
						_user.Write([]byte("\r\033[K" + message + fmt.Sprintf("[%s][%s]:", time.Now().Format("2006-01-02 15:04:05"), private)))
					}
					users[name].Write([]byte(fmt.Sprintf("[%s][%s]:", time.Now().Format("2006-01-02 15:04:05"), name)))
				} else if text == ">list_user" {
					listUsers(user)
					user.Write([]byte(fmt.Sprintf("[%s][%s]:", time.Now().Format("2006-01-02 15:04:05"), name)))
				} else if text == ">clear" {
					clearScreen(user)
					user.Write([]byte(fmt.Sprintf("[%s][%s]:", time.Now().Format("2006-01-02 15:04:05"), name)))
				} else {
					message = fmt.Sprintf("[%s][%s]:", time.Now().Format("2006-01-02 15:04:05"), name) + text + "\n"
					mutex.Lock()
					broadcast(message, name, true)
					mutex.Unlock()
				}
			}
		}
		broadcast(fmt.Sprintf("👋 %s has left our chat ...\n", name), name, false)
		delete(users, name)
	}
}


lib/utils.go
package netcat

import (
	"fmt"
	"net"
	"os"
	"os/exec"
	"strings"
	"sync"
	"time"
)

var (
	users          = map[string]net.Conn{}
	messages       string
	mutex          sync.Mutex
	LogFile        *os.File
	MaxConnections = 10
	emojiMap       = map[string]string{
		":)":  "\U0001F642", // slightly_smiling_face
		":(":  "\U0001F641", // slightly_frowning_face
		";)":  "\U0001F609", // wink
		"<3":  "\U0001F496", // red heart // yellow heart
		":D":  "\U0001F600", // grinning face
		":P":  "\U0001F61B", // face with tongue
		":O":  "\U0001F62E", // face with open mouth
		":'(": "\U0001F622", // crying face
		":/":  "\U0001F615", // confused face
		":*":  "\U0001F618", // face throwing a kiss
		";(":  "\U0001F622", // crying face
	}
)

// broadcast sends the given message to all connected users except the one with the specified name.
// It also logs the message to the LogFile if the lineCount is less than MaxLines.
// Parameters:
//   - message: The message to be broadcasted to the users.
//   - name: The name of the user who sent the message, used to exclude them from receiving the message.
//
// Goals:
//   - Send the message to all connected users.
//   - Log the message to the LogFile if the lineCount is less than MaxLines.
func broadcast(message, name string, mustLog bool) {
	if mustLog {
		_, err := LogFile.WriteString(message)
		if err != nil {
			fmt.Println("❌ [ERROR]: Cannot write on file " + err.Error())
		}
	}
	for i := range users {
		if i != name {
			users[i].Write([]byte("\r\033[K" + message))
		}
		users[i].Write([]byte(fmt.Sprintf("[%s][%s]:", time.Now().Format("2006-01-02 15:04:05"), i)))
	}
	if mustLog {
		messages += message
	}
}

// clearScreen clears the terminal screen for the given connection.
// It uses the "clear" command to clear the screen.
// Parameters:
//   - conn: A net.Conn representing the connection to the user's terminal.
func clearScreen(conn net.Conn) {
	cmd := exec.Command("clear")
	cmd.Stdout = conn
	cmd.Run()
}

// listUsers sends a list of users currently in the chat to the specified connection.
// Parameters:
//   - conn: A net.Conn representing the connection to the user.
func listUsers(conn net.Conn) {
	userList := "Users in chat:\n"
	for username := range users {
		userList += "- " + username + "\n"
	}
	conn.Write([]byte(userList))
}

// replaceEmojis replaces emoji shortcuts in the given text with their corresponding emoji characters.
// Parameters:
//   - text: A string containing text with emoji shortcuts to be replaced.
//
// Returns:
//   - A string with emoji shortcuts replaced by their corresponding emoji characters.
func replaceEmojis(text string) string {
	for k, v := range emojiMap {
		text = strings.Replace(text, k, v, -1)
	}
	return text
}


internal/client.go
package internal

//
import (
	"bufio"
	"fmt"
	"io"
	"net"
	"strings"
	"sync"
)

// checking if a name is provide and return a true that will allow access to the chat
func accessChat(connect net.Conn, connectAddr net.Addr, group *sync.Map, history string) (bool, string) {
	name := ""
	count := 0

	for name == "" || len(name) > 20 {
		if count == 3 {
			connect.Write([]byte("Please provide a non-empty name\n"))
		} else if count > 10 {
			connect.Write([]byte("You're a funny One !\n"))
		} else if count >= 50 {

		}
		connect.Write([]byte("[ENTER YOUR NAME]: "))
		count++
		naming, err := bufio.NewReader(connect).ReadString('\n')
		if err != nil {
			if err == io.EOF {
				// leaveMessage(connect)
				Addr := connect.RemoteAddr().String()
				fmt.Println("Joining Cancelled ... from : " + Addr)
				return false, naming
			} else {
				fmt.Println(err)
				return false, naming
			}
		}

		name = strings.TrimSuffix(naming, "\n")
		if len(name) > 20 {
			connect.Write([]byte("Please Choose a smaller pseudo name"))
		}
	}
	// message := string(historyServing())
	joinMessage(connect, group, name)
	message := string(historyServing(history)) + label(name)
	connect.Write([]byte(message))
	// connect.Write([]byte(message))
	return true, name
}
func connectionHandler(connection net.Conn, name string, group *sync.Map, history string) {

	// defer closeConnection(connection, group)

	for {
		chatData, err := bufio.NewReader(connection).ReadString('\n')
		if err != nil {
			if err == io.EOF {
				// fmt.Println(name + "Has LEFT !!!")
				closeConnection(connection, group, name, history)
				group.Delete(name)
				return
			} else {
				fmt.Println(err)
				return
			}
		}
		// if INFOS != "" {
		// 	logHistory(INFOS, history)
		// 	INFOS = ""
		// }

		message := label(name) + chatData

		logHistory(message, history)

		// var point CLIENTS
		// func (m *Map)
		// f := Load(key any) (value any, ok bool)
		// code := group
		group.Range(func(key, value any) bool {
			// val, ok := group.Load(key)
			// if ok {
			// messageBrodcast(connection, value.(net.Conn), message)
			// point := value.(net.Conn)
			if value.(net.Conn) != connection && chatData != "\n" {
				infos := "\n" + label(name) + chatData + label(fmt.Sprintf("%s", key))
				value.(net.Conn).Write([]byte(infos))
				// value.(net.Conn).Write([]byte("YayY"))
				// } else {
				// }
			}
			return true
		})
		connection.Write([]byte(label(name)))

		if strings.TrimSpace(string(chatData)) == "STOP" || strings.TrimSpace(string(chatData)) == "EXIT" {
			fmt.Print(name + "Exiting Chat ... !")
			connection.Write([]byte("You've successfully logout !"))
			closeConnection(connection, group, name, history)
			// group.Delete(name)
			// connection.Close()
		}
	}
}


internal/server.go
package internal

import (
	"fmt"
	"log"
	"net"
	"os"
	"sync"
	"time"
)

var (
	MAX_CLIENTS = 10
	INFOS       = ""
	HISTORY     = "data/"
	// NETWORK     []CLIENTS
)

type CLIENTS struct {
	history string
	group   sync.Map
}

func Server() {

	PORT := ":"
	if len(os.Args) == 1 {
		PORT += "8989"
	} else {
		PORT += os.Args[1]
	}

	listener, err := net.Listen("tcp", PORT)
	if err != nil {
		fmt.Println(err)
		fmt.Println("Choose between 1024 & 65535")
		return
	}
	defer listener.Close()

	err = os.RemoveAll("data")
	// err = os.Remove("data")
	if err != nil {
		fmt.Println(err)
	}

	if err = os.Mkdir("data", os.ModePerm); err != nil {
		log.Fatal(err)
	}
	fmt.Println("chatLogs directory successfully initialized !")

	fmt.Println("Server Ready on: ", PORT)

addGroup:

	// var connMap map[string]net.Conn
	// HISTORY = "data/history" +
	var groupie CLIENTS
	// NETWORK = append(NETWORK, groupie)
	var connectMap = &sync.Map{}

	groupie.history = handlingLogs()

	for {
		// group, right := chooseChat(NETWORK)
		if checkGroup(&groupie.group) {
			goto addGroup
		}
		// else {
		// groupie = group
		// }
		connection, err := listener.Accept()
		connectionAddr := listener.Addr()
		if err != nil {
			fmt.Println(err)
			continue
		}
		connection.Write(greeting())

		// name, _ := Client(&connection)
		access, name := accessChat(connection, connectionAddr, connectMap, groupie.history)
		// fmt.Println(name + " joined")
		if access {
			groupie.group = *connectMap
			connectMap.Store(name, connection)
			// connection.Write([]byte(label(name)))
			// connMap[name] = connection
			// fmt.Println(connectMap)
			go connectionHandler(connection, name, connectMap, groupie.history)
		}

	}

}

// func chooseChat(rooms []CLIENTS) (CLIENTS, bool) {
// 	empty := CLIENTS{}
// 	for _, chat := range rooms {
// 		if !checkGroup(&chat.group) {
// 			return chat, true
// 		} else {
// 			continue
// 		}
// 	}
// 	return empty, false
// }

func checkGroup(group *sync.Map) bool {
	count := 0
	group.Range(func(key, value any) bool {
		count++
		return true
	})
	if count == MAX_CLIENTS {
		return true
	} else {
		return false
	}
}

func messageBroadcast(group *sync.Map, connect net.Conn) {

}

// func newMessage(msg string, connection net.Conn) message {

// 	address := connection.RemoteAddr().String()
// 	return message{
// 		text:    address + msg,
// 		address: address,
// 	}
// }

func greeting() []byte {
	file, err := os.ReadFile("assets/welcome.txt")
	if err != nil {
		fmt.Println("Argh Welcome Message Missing !!!")
	}
	return file
}

func historyServing(history string) []byte {
	file, err := os.ReadFile(history)
	if err != nil {
		fmt.Println("Sorry chat history unavailable !")
		return []byte("Sorry chat history unavailable !\n")
	}
	if file == nil {
		return []byte("there is no chat history... \n")
	}
	return file
}

func label(name string) string {
	timing := timer()
	infos := timing + "[" + name + "]:" // + string(chatData)
	// fmt.Print(infos)
	return infos
}

func timer() string {
	t := time.Now()
	formatted := fmt.Sprintf("[%d-%02d-%02d %02d:%02d:%02d]",
		t.Year(), t.Month(), t.Day(),
		t.Hour(), t.Minute(), t.Second())
	return formatted
}

func closeConnection(conn net.Conn, group *sync.Map, name string, history string) {
	conn.Close()
	group.Delete(name)
	// message := name + " has left our chat... \n"
	// logHistory(message, history)
	// logHistory(message, history)
	// err := os.WriteFile(HISTORY, []byte(message), 0666)
	// if err != nil {
	// 	fmt.Println("Error logging chat history ...", err)
	// }
	group.Range(func(key, value any) bool {
		if key != name {
			value.(net.Conn).Write([]byte("\n" + name + " has left our chat... \n"))
			value.(net.Conn).Write([]byte(label(fmt.Sprint(key))))
		}
		return true
	})
}

func joinMessage(conn net.Conn, group *sync.Map, name string) {

	// "\r\033[K"
	message := name + " has joined our chat... "
	// INFOS += message + "\n"
	// err := os.WriteFile("/data/hitory.txt", []byte(message), 0666)
	// if err != nil {
	// }
	// var container map[string]struct{}
	// container = map[string]struct{}{}
	group.Range(func(key, value interface{}) bool {
		msg := "\n" + message + "\n" + label(fmt.Sprint(key))
		if key == name {
		} else {
			connect := value.(net.Conn)
			connect.Write([]byte(msg))
		}
		return true
	})
}

func logHistory(message string, history string) {
	file, err := os.OpenFile(history, os.O_APPEND|os.O_WRONLY, 0644)
	if err != nil {
		fmt.Println("Error with chat history file !", err)
		log.Println(err)
	}
	defer file.Close()
	if _, err := file.WriteString(message); err != nil {
		fmt.Println("Error logging chat history !", err)
		// log.Fatal(err)
	}
}

func handlingLogs() string {
	moment := time.Now()
	HISTORY += moment.String() + ".txt"
	err := os.WriteFile(HISTORY, []byte(""), 0666)
	if err != nil {
		fmt.Println("Error creating chat history ...", err)
	}
	return HISTORY
}
